### PR TITLE
Offline Micro Gateway with API key

### PR DESCRIPTION
**Pull request by sabeena1217**
### Proposed changes in this pull request

- Enables the offline micro api gateway, with the API key
 (can run the gateway without any connection with the APIM core)

### When should this PR be merged

ASAP

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/auth/Authentication.bal [hunk 1]
```diff
@@ -210,6 +210,8 @@ function validateSubscription (string apiContext, string version, dto:Introspect
 }
 function validateResource (string apiContext, string apiVersion, string uriTemplate, string verb) (dto:ResourceDto) {
     dto:ResourceDto resourceDto = holder:getFromResourceCache(apiContext, apiVersion, uriTemplate, verb);
+    system:println("resourceDto :");
+    system:println(resourceDto);
```
Comment (tgtshanika) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/auth/Authentication.bal]: Remove unwanted console logs

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/microgateway/microUtil.bal [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    //to find out the resources of that particular api
+
+    json apiList = readFromJSONFile();
+    json resources;
+    int i = 0;
+    int count = jsons:getInt(apiList, "$.length()");
+    string context;
+    string version;
+
+    while (i<count) {
```
Comment (tgtshanika) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/microgateway/microUtil.bal]: Formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/resources/analytics-api.yaml [hunk 3]
```diff
@@ -0,0 +1,759 @@
+swagger: '2.0'
```
Comment (pubudu538) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.commons/src/main/resources/analytics-api.yaml]: I guess this file is not needed here. Please check the usability of this file.

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/event/analytics/response/response.bal [hunk 4]
```diff
@@ -45,8 +45,11 @@ function mediate (message m, message res) {
 
     dto:RequestEventDTO requestEventDTO = {};
     requestEventDTO.api = messages:getProperty(m, "api");
+    system:println(messages:getProperty(m, "api"));
```
Comment (pubudu538) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/event/analytics/response/response.bal]: Do we need system logs here?

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/microgateway/microUtil.bal [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+function loadConfigs () {
+
+    string name = system:getEnv(Constants:GW_HOME);
+    files:File t = {path:name + "/microgateway/microConf.json"};
+
+    if (files:exists(t)) {
+        try {
+            //files:File t = {path:name + "/microgateway/microConf.json"};
+            //not working in windows                                     ????????????????
```
Comment (pubudu538) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/microgateway/microUtil.bal]: Let's add these logs in a proper way. Maybe we can use todo and say that we need to fix the following.

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/microgateway/microUtil.bal [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+            files:open(t, "r");
+            var content, n = files:read(t, 100000000);
+            //so there's a limit! only 100000000bytes can be read        ????????????????
+
+            string strConf = blobs:toString(content, "utf-8");
+            json conf = util:parse(strConf);
+            dto:GatewayConfDTO gatewayConfDTO = gatewayUtil:fromJsonToGatewayConfDTO(conf);
+            //var gatewayConfDTO, _ = <dto:GatewayConfDTO>conf.gatewayConf;
```
Comment (pubudu538) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/microgateway/microUtil.bal]: Let's not add the commented code. If you want to add the commented code, better to comment the reasons for having these.

File: gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/throttle/Throttling.bal [hunk 7]
```diff
@@ -73,13 +73,13 @@ function isRequestThrottled( message msg) (boolean){
...
-        setThrottledResponse(msg);
-        return true;
-    }
+    //if (authorizedUser == ""){
+    //    http:setStatusCode( msg, HTTP_UNAUTHORIZED);
+    //    messages:setProperty(msg, THROTTLED_ERROR_CODE, BLOCKED_ERROR_CODE);
+    //    messages:setProperty(msg, THROTTLED_OUT_REASON, THROTTLE_OUT_REASON_REQUEST_BLOCKED);
+    //    setThrottledResponse(msg);
+    //    return true;
+    //}
```
Comment (ChamNDeSilva) [inline gateway-modules/apim-ballerina-configs/configs/org/wso2/carbon/apimgt/gateway/throttle/Throttling.bal]: Do we need this commented code ?

**Pull request closed without merging**